concepts/eventing-code-guidelines.md

```diff
@@ -0,0 +1,1161 @@
+
+func WithServiceBefore(host, svcName string, apiRule *apigatewayv1alpha1.APIRule) {
+	port := uint32(443)
+	isExternal := true
+	apiRule.Spec.Service = &apigatewayv1alpha1.Service{
+		Name:       &svcName,
+		Port:       &port,
+		Host:       &host,
+		IsExternal: &isExternal,
+	}
+}
+
+apiRule = NewAPIRule(
+	subscriptionWithoutWebhookAuth, 
+	WithPath
+)
+WithServiceBefore(host, svcName, apiRule) // this applies a side effect
+```
+</details>
+
+<details>
+	<summary>Do</summary>
+
+In this example `WithServiceAfter` is used as an argument to `NewAPIRule`.
+
+```go
+func WithServiceAfter(name, host string) APIRuleOption {
+	return func(r *apigatewayv1alpha1.APIRule) {
+		port := uint32(443)
+		isExternal := true
+		r.Spec.Service = &apigatewayv1alpha1.Service{
+			Name:       &name,
+			Port:       &port,
+			Host:       &host,
+			IsExternal: &isExternal,
+		}
+	}
+}
+
+apiRule = NewAPIRule(
+	subscriptionWithoutWebhookAuth,
+	WithPath(),
+	WithServiceAfter(svcName, host), // instead pass it as an argument to NewAPIRule
+)
+```
+</details>
+
+**Example 3**: Side effect as new function
+
+<details>
+  <summary>Don't</summary>
+
+`WithStatusReady` is not used inside a function which is accepting a functional option as input.
+As a reader you would not expect that it performs a `side-effect` on the `apiRuleNew`.
+
+```go
+// source: https://github.com/kyma-project/kyma/blob/a84f76f674babfc63e369bbba44de76499cf475d/components/eventing-controller/controllers/subscription/beb/reconciler_test.go#L592
+getAPIRule(ctx, apiRuleNew).Should(And(
+	HaveNotEmptyHost(),
+	HaveNotEmptyAPIRule(),
+))
+WithStatusReady(apiRuleNew) // <= WithStatusReady not used inside a function which is accepting a functional option as input (e.g. `APIRuleOption`)
+```
+
+</details>
+
+<details>
+  <summary>Do</summary>
+
+In order to make the side-effect more obvious this example uses a function named `MarkReady` instead.
+
+```go
+getAPIRule(ctx, apiRuleNew).Should(And(
+	HaveNotEmptyHost(),
+	HaveNotEmptyAPIRule(),
+))
+MarkReady(apiRuleNew) // instead consider using another function which explicitly states from the name that it applies a side effect
+```
+</details>
+
+## Testing Pyramid in Kyma
+
+The following describes the different testing levels for Kyma. These levels are usually seen as a [pyramid](https://en.wikipedia.org/wiki/Test_automation#Testing_at_different_levels):
+- at the `lower end` of the pyramid you should have the `most` test cases because they are usually very `fast`, however you are limited in what you can test because external dependencies are `mocked`. Unit tests are an example of this.
+- at the `higher` end of the pyramid there are usually `less` tests because these tests take `longer`. However these tests are a more realistic scenario of how the end user will use our product. E2E tests are an example of this.
```

Comment:
```suggestion
- at the `higher` end of the pyramid, there are usually `less` tests because these tests take `longer`. However, these tests are a more realistic scenario of how the end-user will use our product. E2E tests are an example of this.
```

Reply:
I agree with the commas but [end user](https://www.merriam-webster.com/dictionary/end%20user) needs no hyphen :D 
(in general, [noun compounds](https://help.sap.com/viewer/DRAFT/e33c591ae4494a659a3f5f983c9d1161/PROD/en-US/b703f0cb978f49aeb3e35ebb4e051a4b.html) usually don't need a hyphen, adjective compounds often do. Probably just to keep things interesting ¯\_(ツ)_/¯ )

Reply:
@k15r @NHingerl Commas found its way inside the sentence. I assume through a review suggestion from @NHingerl.
End user is without hyphen.